src/lib/zork/zork.ts

```diff
@@ -0,0 +1,378 @@
+      // TODO: backpressure!
+      const BUFFTHRESHOLD = 16000000;  // 16 megabytes
+      if (channel.bufferedAmount < BUFFTHRESHOLD) {
+        channel.send(bytes);
+      } else {
+        console.warn('channel congested, dropping bytes')
+      }
+    });
+    socksSession.onDisconnect(() => {
+      console.info(`[get] ${sessionId}: socks session disconnected`);
+    });
+    channel.onclose = () => {
+      console.info(`[get] ${sessionId}: channel closed (giver side)`);
+    };
+  };
+};
+
+const cmdHandlerByCmdId: {[cmdId: string]: (ctx: Context, cmd?: ParsedCmd) => void} = {
+  'ping': handleCmdPing,
+  'xyzzy': handleCmdXyzzy,
+  'version': handleCmdVersion,
+  'quit': handleCmdQuit,
+  'getters': handleCmdGetters,
+  'transform': handleCmdTransform,
+  'give': handleCmdGive,
+  'get': handleCmdGet
+};
+
+const parseCmd = (cmdline: string) : ParsedCmd => {
+  const tokens = cmdline.split(/\W+/);
+  const cmdId = tokens[0].toLowerCase();
+  const parsed = {id: cmdId, tokens: tokens, source: cmdline};
+  return parsed;
+};
+
+
+// Message handlers:
+
+const handleMsgForModeGive = (ctx: Context, msg: string) => {
+  const parsed = JSON.parse(msg);
+  if (parsed.sdp) {
+    console.info(`[give] got sdp`);//: ${msg}`);
+    ctx.peerConn.setRemoteDescription(parsed);
+  } else {
+    console.info(`[give] parsed.sdp missing, ignoring msg`);//: ${msg}`);
+  }
+};
+
+const handleMsgForModeGet = (ctx: Context, msg: string) => {
+  const parsed = JSON.parse(msg);
+  if (parsed.type === 'icecandidate') {
+    console.info(`[get] icecandidate from client ${ctx.clientId}`);//: ${msg}`);
+    ctx.peerConn.addIceCandidate(parsed.candidate);
+  } else if (parsed.type === 'offer') {
+    console.info(`[get] got offer`);//: ${msg}`);
+    ctx.peerConn.setRemoteDescription(parsed);
+    ctx.peerConn.createAnswer((answer: any) => {
+      const answerJson = JSON.stringify(answer);
+      console.info(`[get] created answer`);//: ${answerJson}`);
+      ctx.peerConn.setLocalDescription(answer);
+      ctx.socket.write(`${JSON.stringify(answer)}\n`);
+    }, console.error);
+  } else {
+    console.error(`[get] unexpected msg: ${msg}`);
+  }
+};
+
+const msgHandlerByMode: {[mode: string]: (ctx: Context, msg: string) => void} = {
+  'give': handleMsgForModeGive,
+  'get': handleMsgForModeGet
+};
+
+const handleMsg = (ctx: Context, msg: string) => {
+  if (ctx.mode) {
+    // Already in 'give' or 'get' mode. Dispatch to corresponding message handler.
+    const msgHandler = msgHandlerByMode[ctx.mode];
+    if (msgHandler) {
+      msgHandler(ctx, msg);
+    } else {
+      console.error(`no message handler for mode: ${ctx.mode}`);
+    }
+  } else {
+    // Not yet in 'give' or 'get' mode. Treat msg as command.
+    const cmd = parseCmd(msg);
+    const cmdHandler = cmdHandlerByCmdId[cmd.id] || handleCmdInvalid;
+    cmdHandler(ctx, cmd);
+  }
+};
+
+const zorkServer = net.createServer((client) => {
```

Comment:
Just happened to find the term for this style ("fluent") as well as a discussion of its tradeoffs on Wikipedia: https://en.wikipedia.org/wiki/Fluent_interface#Problems
#wikivictories